xarray/core/formatting.py

```diff
@@ -222,38 +222,28 @@ def _summarize_coord_multiindex(coord, col_width, marker):
 def _summarize_coord_levels(coord, col_width, marker=u'-'):
     relevant_coord = coord[:30]
     return u'\n'.join(
-        [_summarize_var_or_coord(lname,
-                                 relevant_coord.get_level_variable(lname),
-                                 col_width, marker=marker)
+        [summarize_variable(lname,
+                            relevant_coord.get_level_variable(lname),
+                            col_width, marker=marker)
          for lname in coord.level_names])
 
 
-def _not_remote(var):
-    """Helper function to identify if array is positively identifiable as
-    coming from a remote source.
-    """
-    source = var.encoding.get('source')
-    if source and source.startswith('http') and not var._in_memory:
-        return False
-    return True
-
-
-def summarize_var(name, var, col_width):
-    show_values = _not_remote(var)
-    return _summarize_var_or_coord(name, var, col_width, show_values)
+def summarize_datavar(name, var, col_width):
+    show_values = var._in_memory
```

Comment:
Our current heuristic uses the `_not_remote()` helper function, so it doesn't display arrays loaded over a network (via opendap), which can often be quite slow. But it does display a summary of values from netCDF files on disk, which I do think is generally helpful and for which I haven't noticed any performance issues.

Based on the current definition of `_in_memory`, we wouldn't display any of these arrays:
https://github.com/pydata/xarray/blob/4a15cfa6f3e680af6969f3e2c6a375e2cc8c3288/xarray/core/variable.py#L285-L289

So instead of using `_in_memory`, I would suggest something like `_not_remote(var) and not isinstance(var._data, dask_array_type)` as the condition for showing values.

Reply:
@shoyer loading a NetCDF variable from disk every time you do __repr__ is a terrible idea if that variable has been compressed withouth chunking. If the variable is a single block of 100MB of zlib-compressed data, you will have to read it and decompress it every time.

Reply:
@shoyer also, your netcdf array might be sitting on a network file system on the opposite side of a narrowband VPN.

Reply:
That's certainly possible, but in my experience very few people writes 100MB chunks -- those are very large.

Let's summarize our options:
1. Always show a preview of data from netCDF files with `Dataset.__repr__`
2. Never show a preview for data if it isn't already in memory
3. Show a preview depending on a global option (with default choice TBD).

Reasons to show data from disk in `__repr__`:
- It's what we've always done.
- "Most" of the time it's fast and convenient.
- It provides a good experience for new users, who don't need to hunt for a separate `preview()` or `load()` command to see what's in a Dataset. You can simply print it at a console.

Reasons not to show data from disk in `__repr__`:
- IO can be slow/expensive, especially if compression or networks are involved.
- Heuristics to detect expensive IO are unreliable and somewhat distasteful.

Maybe we should solicit a few more opinions here before we change the default behavior?

Another possibility is to try loading data in a separate thread and timeout if it takes too long (say more than 0.5 seconds), but that might open up it's own set of performance issues (it's not easy to kill a thread, short of terminating a process).

Reply:
I think my vote would be to only print a preview of data that is in memory. For my uses, I typically have fill values in the first 10-20 data points so the previous `__repr__` didn't give me any information.

@pydata/xarray - anyone else have thoughts on this?

Reply:
I am also for only showing only data in memory and never read from disk for a ``__repr__``, for the same reasons. Looking forward to see this PR merged!

Reply:
@shoyer - do we have results from your google poll on this issue yet?

Reply:
Sounds like I was wrong -- the consensus is pretty clear that we should go ahead with this

<img width="735" alt="screen shot 2017-09-20 at 12 22 32 pm" src="https://user-images.githubusercontent.com/1217238/30663137-80bc3896-9dfe-11e7-9929-2b3d0f94f7fd.png">
<img width="724" alt="screen shot 2017-09-20 at 12 22 38 pm" src="https://user-images.githubusercontent.com/1217238/30663155-9330c366-9dfe-11e7-9a0c-98cd3d46c18a.png">
<img width="704" alt="screen shot 2017-09-20 at 12 22 41 pm" src="https://user-images.githubusercontent.com/1217238/30663158-94d9e634-9dfe-11e7-8969-d6694f0e4d97.png">
<img width="736" alt="screen shot 2017-09-20 at 12 22 48 pm" src="https://user-images.githubusercontent.com/1217238/30663161-9769f088-9dfe-11e7-8d66-9ebb376ef483.png">

Reply:
I'm not sure this sample size is going to give us statistically significant results but I'm glad to see @delgadom and I are in agreement.

@crusaderky - are you up for implementing this?

Reply:
I think the current implementation (in this PR) is actually already correct.

Reply:
Yep - data is eagerly loaded from disk only for index coords on \_\_init\_\_ now.